docs/tutorials/essentials/part-4-using-data.md

```diff
@@ -675,169 +711,514 @@ Typically, social media feeds show the newest posts first, and you scroll down t
-        }
-        // highlight-end
-      >
-        {emoji} {post.reactions[name]}
-      </button>
-    )
-  })
+export const ReactionButtons = ({ post }: ReactionButtonsProps) => {
+  const dispatch = useAppDispatch()
+
+  const reactionButtons = Object.entries(reactionEmoji).map(
+    ([stringName, emoji]) => {
+      // Ensure TS knows this is a _specific_ string type
+      const reaction = stringName as ReactionName
+      return (
+        <button
+          key={reaction}
+          type="button"
+          className="muted-button reaction-button"
+          onClick={() => dispatch(reactionAdded({ postId: post.id, reaction }))}
+        >
+          {emoji} {post.reactions[reaction]}
+        </button>
+      )
+    }
+  )
 
   return <div>{reactionButtons}</div>
 }
 ```
 
-Now, every time we click a reaction button, the counter should increment. If we browse around to different parts of the app, we should see the correct counter values displayed any time we look at this post, even if we click a reaction button in the `<PostsList>` and then look at the post by itself on the `<SinglePostPage>`.
+Now, every time we click a reaction button, the counter for that reaction should increment. If we browse around to different parts of the app, we should see the correct counter values displayed any time we look at this post, even if we click a reaction button in the `<PostsList>` and then look at the post by itself on the `<SinglePostPage>`. This is because each component is reading the same post data from the Redux store.
+
+## Adding User Login
+
+We've got one more feature to add in this section.
+
+Right now, we just select which user is writing each post in the `<AddPostForm>`. To add a bit more realism, we ought to have the user log in to the application, so that we already know who is writing the posts (and be useful for other features later).
+
+Since this is a small example app, **we aren't going to implement any _real_ authentication checks (and the point here is to learn how to use Redux features, not how to actually implement real auth)**. Instead, we'll just show a list of user names and let the actual user select one of them.
+
+For this example, we'll just add an `auth` slice that tracks `state.auth.username` so we know who the user is. Then, we can use that information whenever they add a post to automatically add the right user ID to the post.
+
+### Adding an Auth Slice
+
+The first step is to create the `authSlice` and add it to the store. This is the same pattern we've seen already - define the initial state, write the slice with a couple of reducers to handle updates for login and logout, and add the slice reducer to the store.
+
+In this case, our auth state is really just the current logged-in username, and we'll reset it to `null` if they log out.
+
+```ts title="features/auth/authSlice.ts"
+import { createSlice, PayloadAction } from '@reduxjs/toolkit'
```

Comment:
```suggestion
import type { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from '@reduxjs/toolkit'
```

Reply:
I may apply the `type` import change near the end, but I'm honestly not worried about that for now